### PR TITLE
Add AQHI loader

### PR DESCRIPTION
**Pull request by flaframboise**
I added AQHI data to GeoMet OGC-API. This includes:

- Added new loader for Forecast and observations AQHI data, as well as AQHI regions
- Added new GeoJSON (`aqhi_community.geojson`) for `aqhi_regions` ES index
- Added webpage for related collection
- Added cleanup cron for AQHI (keep last 3 days of data)

### CODE REVIEW

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+                features = data['features']
+            elif self.type == "observations":
+                features = [data]
+
+        for feature in features:
+            # set document id
+            feature['id'] = feature['ID']
+
+            # clean out unnecessery properties
+            feature.pop('ID')
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:313]: safeguard with `feature.pop('ID', None)`

Comment (tomkralidis) [reply msc_pygeoapi/loader/aqhi_realtime.py:313]: Actually:

```python
feature['id'] = feature.pop('ID', None)
```

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+        # generate geojson features
+        package = self.generate_geojson_features()
+        self.conn.submit_elastic_package(package, request_size=80000)
+
+        return True
+
+
+@click.group()
+def aqhi_realtime():
+    """Manages AQHI indices"""
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:355]: `Manages AQHI index`

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+        for root, dirs, files in os.walk(directory):
+            for f in [file for file in files if file.endswith('.json')]:
+                files_to_process.append(os.path.join(root, f))
+        files_to_process.sort(key=os.path.getmtime)
+
+    for file_to_process in files_to_process:
+        loader = AqhiRealtimeLoader(conn_config)
+        result = loader.load_data(file_to_process)
+        if result:
+            click.echo(
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:389]: Change to:

```python
if not result:
             click.echo('features not generated')
```

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+@click.command()
+@click.pass_context
+@cli_options.OPTION_DAYS(
+    default=DAYS_TO_KEEP,
+    help='Delete indexes older than n days (default={})'.format(DAYS_TO_KEEP),
+)
+@cli_options.OPTION_DATASET(
+    help='AQHI dataset indexes to delete.',
+    type=click.Choice(['all', 'forecasts', 'observations']),
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:430]: Aren't we managing a single index (`aqhi-realtime`) via pygeoapi?

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+@cli_options.OPTION_ES_IGNORE_CERTS()
+@cli_options.OPTION_YES(prompt='Are you sure you want to delete old indexes?')
+def clean_indexes(ctx, days, dataset, es, username, password, ignore_certs):
+    """Delete old AQHI realtime indexes older than n days"""
+
+    conn_config = configure_es_connection(es, username, password, ignore_certs)
+    conn = ElasticsearchConnector(conn_config)
+
+    if dataset == 'all':
+        indexes_to_fetch = '{}*'.format(INDEX_BASENAME.format('*'))
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:444]: Check/ensure index patterns are still valid given the `INDEX_BASENAME` value update.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+@click.pass_context
+@cli_options.OPTION_DATASET(
+    help='AQHI dataset indexes to delete.',
+    type=click.Choice(['all', 'forecasts', 'observations', 'regions']),
+)
+@cli_options.OPTION_ELASTICSEARCH()
+@cli_options.OPTION_ES_USERNAME()
+@cli_options.OPTION_ES_PASSWORD()
+@cli_options.OPTION_ES_IGNORE_CERTS()
+def delete_indexes(ctx, dataset, es, username, password, ignore_certs):
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:469]: See https://github.com/ECCC-MSC/msc-pygeoapi/blob/master/msc_pygeoapi/loader/bulletins_realtime.py#L207-L233 for example (which also includes index template deletion).

File: msc_pygeoapi/plugin.py [hunk 7]
```diff
@@ -66,6 +66,10 @@
             'filename_pattern': 'observations/swob-ml',
             'handler': 'msc_pygeoapi.loader.swob_realtime.SWOBRealtimeLoader'  # noqa
         },
+        'aqhi_realtime': {
+            'filename_pattern': 'air_quality/aqhi',
+            'handler': 'msc_pygeoapi.loader.aqhi_realtime.AqhiRealtimeLoader'  # noqa
+        },
```
Comment (tomkralidis) [inline msc_pygeoapi/plugin.py:72]: Remove trailing comma.

**Review by tomkralidis — CHANGES_REQUESTED** [review]

File: deploy/default/msc-pygeoapi-config.yml [hunk 8]
```diff
@@ -947,6 +947,38 @@ resources:
...
+              rel: canonical
+              title: Données et produits d'observations et prévisions qui sont générés pour le programme Cote Air Santé (CAS)
+              href: https://eccc-msc.github.io/open-data/msc-data/aqhi/readme_aqhi_fr/
+              hreflang: fr-CA
+        extents:
+            spatial:
+                bbox: [-141, 41.68, -52.62, 83.11]
+                crs: http://www.opengis.net/def/crs/OGC/1.3/CRS84
+            temporal:
+                begin: 2013-01-01
```
[the remark this answers is not in the document]
Comment (flaframboise) [reply deploy/default/msc-pygeoapi-config.yml:973]: Will do! However, I see that other time formats are YYYY-MM-DD too...do you think I should change them while I'm at it/ create a new ticket

[the remark this answers is not in the document]
Comment (tomkralidis) [reply deploy/default/msc-pygeoapi-config.yml:973]: Only collections with a provider `time_field` specified need an RFC3339 notation, else the value is for metadata purpose.  Suggest to leave the others for now.

File: msc_pygeoapi/loader/__init__.py [hunk 9]
```diff
@@ -46,6 +46,7 @@
     from msc_pygeoapi.loader.swob_realtime import swob_realtime
     from msc_pygeoapi.loader.ltce import ltce
     from msc_pygeoapi.loader.climate_archive import climate_archive
+    from msc_pygeoapi.loader.aqhi_realtime import aqhi_realtime
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/__init__.py:49]: In the header of this file, add your Author and Copyright.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
+# =================================================================
+#
+# Author: Etienne Pelletier <etienne.pelletier@canada.ca>
+#         Felix Laframboise <felix.laframboise@canada.ca>
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:4]: Add empty line between authors and copyright.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+# Copyright (c) 2020 Etienne Pelletier
+#               2020 Felix Laframboise
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:6]: Change year

File: msc_pygeoapi/plugin.py [hunk 12]
```diff
@@ -66,6 +66,10 @@
             'filename_pattern': 'observations/swob-ml',
             'handler': 'msc_pygeoapi.loader.swob_realtime.SWOBRealtimeLoader'  # noqa
         },
+        'aqhi_realtime': {
```
Comment (tomkralidis) [inline msc_pygeoapi/plugin.py:69]: In the header of this file, add your Author and Copyright.

File: tests/test_aqhi_realtime.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
+# =================================================================
+#
+# Author: Thinesh Sornalingam <thinesh.sornalingam@canada.ca>,
```
Comment (tomkralidis) [inline tests/test_aqhi_realtime.py:3]: Author and copyright should be only yourself.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+                        'type': 'text',
+                        'fields': {'raw': {'type': 'keyword'}},
+                    },
+                    'hourly_forecast_utc': {
+                        'type': 'date',
+                        'format': 'strict_date_time_no_millis',
+                    },
+                    'aqhi': {'type': 'byte'},
+                }
+            },
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:105]: Remove trailing comma.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+                        'type': 'text',
+                        'fields': {'raw': {'type': 'keyword'}},
+                    },
+                    'datetime_text_fr': {
+                        'type': 'text',
+                        'fields': {'raw': {'type': 'keyword'}},
+                    },
+                    'aqhi': {'type': 'float'},
+                }
+            },
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:147]: Remove trailing comma.

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+        }
+    },
+}
+
+SETTINGS = {
+    'order': 0,
+    'version': 1,
+    'index_patterns': ['{}*'.format(INDEX_BASENAME)],
+    'settings': {'number_of_shards': 1, 'number_of_replicas': 0},
+    'mappings': None,
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:157]: Remove trailing comma

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+                    },
+                    'datetime_text_fr': {
+                        'type': 'text',
+                        'fields': {'raw': {'type': 'keyword'}},
+                    },
+                    'aqhi': {'type': 'float'},
+                }
+            },
+        }
+    },
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:149]: Remove trailing comma

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+)
+
+LOGGER = logging.getLogger(__name__)
+elastic_logger.setLevel(getattr(logging, MSC_PYGEOAPI_LOGGING_LOGLEVEL))
+
+# cleanup settings
+DAYS_TO_KEEP = 3
+
+# index settings
+INDEX_BASENAME = 'aqhi_realtime.{}.'
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:59]: ` INDEX_BASENAME = 'aqhi-realtime-{}.'`

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+        # generate geojson features
+        package = self.generate_geojson_features()
+        self.conn.submit_elastic_package(package, request_size=80000)
+
+        return True
+
+
+@click.group()
+def aqhi_realtime():
+    """Manages AQHI index"""
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:273]: `     """Manages AQHI indexes"""`

File: msc_pygeoapi/loader/aqhi_realtime.py [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+@click.command()
+@click.pass_context
+@cli_options.OPTION_FILE()
+@cli_options.OPTION_DIRECTORY()
+@cli_options.OPTION_ELASTICSEARCH()
+@cli_options.OPTION_ES_USERNAME()
+@cli_options.OPTION_ES_PASSWORD()
+@cli_options.OPTION_ES_IGNORE_CERTS()
+def add(ctx, file_, directory, es, username, password, ignore_certs):
+    """Add aqhi data to Elasticsearch"""
```
Comment (tomkralidis) [inline msc_pygeoapi/loader/aqhi_realtime.py:286]: `"""Adds AQHI data to Elasticsearch"""`

File: tests/test_aqhi_realtime.py [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+import pytest
+import requests
+
+
+@pytest.fixture()
+def url(pytestconfig):
+    return pytestconfig.getoption('url')
+
+
+@pytest.mark.parametrize('aqhi', [
```
Comment (tomkralidis) [inline tests/test_aqhi_realtime.py:46]: Are these being used?

File: tests/test_aqhi_realtime.py [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# =================================================================
+
```
Comment (tomkralidis) [inline tests/test_aqhi_realtime.py:33]: Divide this test file into two core testing functions:
- test that the test AQHI data is correctly converted into GeoJSON (`test_loader`)
- the API requests (`test_api`)

**Review by tomkralidis — CHANGES_REQUESTED** [review]

File: msc_pygeoapi/loader/aqhi_realtime.py (hunk 18 above)
Comment (flaframboise) [reply msc_pygeoapi/loader/aqhi_realtime.py:59]: Is it not supposed to be `INDEX_BASENAME = 'aqhi-{}-realtime.'` following comments above?

Comment (tomkralidis) [reply msc_pygeoapi/loader/aqhi_realtime.py:59]: Ah, yes, `aqhi-{}-realtime.'`

File: tests/test_aqhi_realtime.py (hunk 13 above)
Comment (tomkralidis) [reply tests/test_aqhi_realtime.py:3]: Author and copyright should be only yourself.

**Review by tomkralidis — APPROVED** [review]

**Pull request merged**